apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,510 @@
+
+  switch (type) {
+    case VideoGridAction.ResetVideoGridState: {
+      return initialState;
+    }
+    case VideoGridAction.UpdateAttendeeStates: {
+      const { roster } = payload;
+
+      // Remove attendee that left the meeting
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (!isContentShare(attendeeId) && !(attendeeId in roster)) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Add attendee that joined the meeting
+      for (const attendeeId of Object.keys(roster)) {
+        const name = roster[attendeeId]?.name || '';
+
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].name = name;
+        } else {
+          attendeeStates[attendeeId] = {
+            attendeeId,
+            name,
+            videoEnabled: false,
+            bandwidthConstrained: false,
+          } as AttendeeState;
+        }
+      }
+
+      // Ensure the state of `videoEnabled` in the racing condition of UpdateAttendeeStates and UpdateVideoSources
+      // To do: Merge all the update actions to one
+      for (const attendeeId of videoSourceState.cameraSources) {
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].videoEnabled = true;
+        }
+      }
+
+      return {
+        ...state,
+        attendeeStates,
+      };
+    }
+    case VideoGridAction.UpdateVideoSources: {
+      const { videoSources } = payload as { videoSources: VideoSource[] };
+      const cameraSources: string[] = [];
+      const videoSourceIds = videoSources.map(
+        videoSource => videoSource.attendee.attendeeId
+      );
+
+      // Reset the `videoEnabled` of all attendeeStates
+      for (const attendee of Object.values(attendeeStates)) {
+        attendee.videoEnabled = false;
+      }
+
+      // Remove content share from attendeeStates,
+      // if content share is not in video sources
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (
+          isContentShare(attendeeId) &&
+          !videoSourceIds.includes(attendeeId)
+        ) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Update the `videoEnabled` according to video sources
+      for (const attendeeId of videoSourceIds) {
+        if (!(attendeeId in attendeeStates)) {
+          if (isContentShare(attendeeId)) {
+            attendeeStates[attendeeId] = {
+              attendeeId,
+              name: 'content share',
+              bandwidthConstrained: false,
+            } as AttendeeState;
+          } else {
+            attendeeStates[attendeeId] = {
+              attendeeId,
+              name: '',
+              bandwidthConstrained: false,
+            } as AttendeeState;
+          }
+        }
+
+        attendeeStates[attendeeId].videoEnabled = true;
+      }
+
+      // Populate the `cameraSources` based on the order of attendeeStates
+      for (const attendee of Object.values(attendeeStates)) {
```

Comment:
Can we avoid this loop again and in the previous `for` itself keep updating the `cameraSources`:
```js
...
attendeeStates[attendeeId].videoEnabled = true;
cameraSources.push(attendeeId);
```

Reply:
I'm not 100% sure, but I feel like the order matters here. The previous based on the order of id in `videoSources`.
This loop based on the order of `attendeeStates` (roster list).
I'm not sure if there's any relation between the order of remote tiles and the order of video preferences.
So have to keep it. If later we found the order doesn't matter here, we could optimize it, and other places.